crates/services/sync/src/service.rs

```diff
@@ -1,61 +1,232 @@
+                // Select over the task runner state, the sync task and the import task.
+                // Match on either the sync task or the import task completing.
+                // If the task runner state changes then we check it below.
+                if let Either::Right((join_handles, _)) =
+                    select(watcher.changed().boxed(), select(sync_task, import_task))
+                        .await
+                {
+                    match join_handles {
+                        // Sync task has completed.
+                        Either::Left((sync_result, import_task)) => match sync_result {
+                            // Sync task has panicked.
+                            Err(e) if e.is_panic() => {
+                                // Kill the import task and resume the panic.
+                                self.kill_switch.kill_all();
+                                let _ = import_task.await;
+                                std::panic::resume_unwind(e.into_panic());
+                            }
+                            _ => unreachable!(
+                                "The sync task has no errors so this can't happen"
+                            ),
+                        },
+                        // Import task has completed.
+                        Either::Right((import_result, sync_task)) => {
+                            match import_result {
+                                Ok(task_result) => return task_result,
+                                Err(e) => {
+                                    // If the import task has panicked then we kill the sync task
+                                    // and resume the panic.
+                                    if e.is_panic() {
+                                        self.kill_switch.kill_all();
+                                        let _ = sync_task.await;
+                                        std::panic::resume_unwind(e.into_panic());
+                                    } else {
+                                        return Err(e.into())
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+            _ => return Ok(false),
         }
-    }
 
-    pub async fn stop(&self) -> Option<JoinHandle<()>> {
-        let join = self.join.lock().take();
-        if join.is_some() {
-            let _ = self.sender.send(SyncMpsc::Stop);
+        // Check if the state of the task runner is still running.
+        if !watcher.borrow().started() {
+            // The task runner has stopped so we kill the sync and import tasks.
+            self.kill_switch.kill_all();
+            if let Some(sync_task) = self.sync_task.take() {
+                let _ = sync_task.await;
+            }
+            if let Some(import_task) = self.import_task.take() {
+                let _ = import_task.await;
+            }
```

Comment:
Could you create a `shutdown` method in the `RunnableTask` that we discussed and do it there, please? In this case, you don't need to `!watcher.borrow().started()`

Reply:
If we don't check the watcher here then there's no way to shutdown the tasks based on the state changing to stopped though. Maybe I don't understand what you mean?

Reply:
It is already checked in the in the `ServiceRunner` 
<img width="676" alt="image" src="https://user-images.githubusercontent.com/18346821/213034905-931f350d-4c72-4e22-bc2a-c5409a558d57.png">